core/src/shred_fetch_stage.rs

```diff
@@ -268,6 +249,51 @@ impl ShredFetchStage {
     }
 }
 
+fn receive_quic_datagrams(
+    quic_endpoint_receiver: Receiver<(Pubkey, SocketAddr, Bytes)>,
+    sender: Sender<PacketBatch>,
+    recycler: PacketBatchRecycler,
+    exit: Arc<AtomicBool>,
+) {
+    const RECV_TIMEOUT: Duration = Duration::from_secs(1);
+    while !exit.load(Ordering::Relaxed) {
+        let entry = match quic_endpoint_receiver.recv_timeout(RECV_TIMEOUT) {
+            Ok(entry) => entry,
+            Err(RecvTimeoutError::Timeout) => continue,
+            Err(RecvTimeoutError::Disconnected) => return,
+        };
+        let entries = std::iter::once(entry).chain(
+            std::iter::repeat_with(|| {
+                quic_endpoint_receiver
+                    .recv_timeout(PACKET_COALESCE_DURATION)
+                    .ok()
+            })
+            .while_some(),
+        );
```

Comment:
Unless I'm missing something obvious, this isn't really a coalesce, since it's going to wait up to `PACKET_COALESCE_DURATION` until it stops receiving, which could go on forever. Probably you want to do a `select!` on some elapsed duration (https://docs.rs/crossbeam-channel/latest/crossbeam_channel/macro.select.html#examples) or a `take_while` and check the elapsed time.

Reply:
Switched to `recv_deadline`.

Even with the old code it would stop at max `PACKETS_PER_BATCH` because it was doing `.zip(packet_batch.iter_mut())`. But I guess `recv_deadline` gives a more accurate coalesce implementation anyways.